check-couchdb-cluster.rb

```diff
@@ -0,0 +1,162 @@
+#! /usr/bin/env ruby
+#
+#  check-couchdb-cluster.rb
+#
+# DESCRIPTION:
+#   Checks the health, size and version of a CouchDB cluster
+#
+# PLATFORMS:
+#   Linux
+#
+# DEPENDENCIES:
+#   gem: sensu-plugin
+#   gem: json
+#   gem: rest-client
+#
+# NOTES
+#   Based on:
+#     https://github.com/sensu-plugins/sensu-plugins-couchbase/blob/master/bin/check-couchbase-cluster.rb
+#
+# LICENSE:
+#   Copyright 2017 Pieter Vogelaar
+#   Released under the same terms as Sensu (the MIT license); see LICENSE
+#   for details.
+#
+
+require 'sensu-plugin/check/cli'
+require 'json'
+require 'rest_client'
+
+# main plugin class
+
+class CouchDBCluster < Sensu::Plugin::Check::CLI
+  option :user,
+         description: 'CouchDB API auth username',
+         short: '-u USERNAME',
+         long: '--user USERNAME'
+
+  option :password,
+         description: 'CouchDB API auth password',
+         short: '-P PASSWORD',
+         long: '--password PASSWORD'
+
+  option :host,
+         description: 'CouchDB host / Load balancer DNS name',
+         short: '-h HOST',
+         long: '--host HOST',
+         default: 'localhost'
+
+  option :port,
+         description: 'CouchDB port',
+         short: '-p PORT',
+         long: '--port PORT',
+         default: 5984
+
+  option :cluster_size,
+         description: 'Cluster size expected',
+         short: '-c CLUSTER_SIZE',
+         long: '--cluster-size CLUSTER_SIZE',
+         proc: proc(&:to_i)
+
+  option :couchdb_version,
+         description: 'CouchDB version expected',
+         short: '-v VERSION',
+         long: '--version VERSION'
+
+  def run
+    begin
+      resource = '/_membership'
+      response = RestClient::Request.execute(
+          method: :get,
+          url: "http://#{config[:host]}:#{config[:port]}/#{resource}",
+          user: config[:user],
+          password: config[:password],
+          headers: { accept: :json, content_type: :json }
+      )
+      result = JSON.parse(response.to_str, symbolize_names: true)
+    rescue Errno::ECONNREFUSED
+      unknown 'Connection refused'
+    rescue RestClient::ResourceNotFound
+      unknown "Resource not found: #{resource}"
+    rescue RestClient::RequestFailed
+      unknown 'Request failed'
+    rescue RestClient::RequestTimeout
+      unknown 'Connection timed out'
+    rescue RestClient::Unauthorized
+      unknown 'Missing or incorrect CouchDB API credentials'
+    rescue JSON::ParserError
+      unknown 'CouchDB API returned invalid JSON'
+    end
+
+    # Get information about each node
+    nodes = {}
+    result[:all_nodes].each do |node|
+      node_parts = node.split('@')
+      nodes[node_parts[1]] = get_node(node_parts[1], config[:port], config[:user], config[:password])
+    end
+
+    # Collect node CouchDB versions
+    node_versions = []
+    nodes.each do |fqdn, node|
+      node_versions << node[:version]
+    end
+
+    # Check CouchDB version consistency
+    if node_versions.uniq.length > 1
```

Comment:
I have not used couchbase in a while but I recall you could do some kind of rolling upgrade for minor version, perhaps we should make this an option that the user can control. Otherwise they will get alerts when they are part of the way through an upgrade. I think there are a few routes:
- expose an option to control the number of allowed versions (for example you might say 2 versions are ok because you might be upgrading but 3 is always wrong)
- expose an option to control what severity of an incident is.
- expose an option to disable the alert but include some output to the OK text.

Reply:
verified that couchdb does support rolling upgrades, ironically I found it here: http://pietervogelaar.nl/rolling-upgrade-elasticsearch-couchbase-and-couchdb